fuzzers/main.py

```diff
@@ -0,0 +1,337 @@
+
+REPO_DIR_PARENT = WORKDIR
+REPO_DIR_NAME = 'fuzzed-nearcore'
+REPO_DIR = REPO_DIR_PARENT / REPO_DIR_NAME
+
+GCS_CREDENTIALS_FILE = config.CONFIG_DIR / 'credentials.json'
+GCS_BUCKET = 'fuzzer'
+
+FUZZ_TIME = prometheus_client.Counter('fuzz_seconds', 'Time spent fuzzing', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CRASHES = prometheus_client.Counter('fuzz_crashes', 'Number of times the fuzzer process crashed (not unique crashes)', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_ARTIFACTS_FOUND = prometheus_client.Counter('fuzz_artifacts_found', 'Number of artifacts found (should be number of unique crashes)', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CORPUS_UPLOADED = prometheus_client.Counter('fuzz_corpus_uploaded', 'Number of elements uploaded to GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CORPUS_DELETED = prometheus_client.Counter('fuzz_corpus_deleted', 'Number of elements deleted from GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+
+def update_repo(branch):
+    if not REPO_DIR.exists():
+        print(f"Doing initial clone of repository {REPO_DIR}")
+        subprocess.check_call(['git', 'clone', REPO_URL, REPO_DIR_NAME], cwd=REPO_DIR_PARENT)
+
+    print(f"Updating to latest commit of branch {branch}")
+    subprocess.check_call(['git', 'fetch', REPO_URL, branch], cwd=REPO_DIR)
+    subprocess.check_call(['git', 'checkout', 'FETCH_HEAD'], cwd=REPO_DIR)
+
+def parse_config():
+    return toml.load(REPO_DIR / 'nightly' / 'fuzz.toml')
+
+FUZZERS = []
+
+def report_artifact(gcs_path, branch, t):
+    # todo: report on zulip with all the details
+    FUZZ_ARTIFACTS_FOUND.labels(branch['name'], t['crate'], t['runner'], t['flags']).inc()
+
+def artifact_deleted(path):
+    raise RuntimeError(f"Deleted artifact {path}, which should have been prevented by ignore_deletions")
+
+def run_fuzzer(branch, t, gcs, exit_event):
+    # First, figure out the current corpus version
+    bucket = gcs.get_bucket(GCS_BUCKET)
+    corpus_vers = bucket.blob('current-corpus').download_as_text().strip()
+
+    date = datetime.datetime.now().strftime('%Y-%m-%d')
+    log_path = pathlib.Path(date) / t['crate'] / t['runner'] / str(uuid.uuid4())
+    log_filename = WORKDIR / 'fuzz-logs' / log_path
+    utils.mkdirs(log_filename.parent)
+    with open(log_filename, 'a') as log_file:
+        current_commit = subprocess.run(['git', 'rev-parse', 'HEAD'], cwd=REPO_DIR, capture_output=True, check=True).stdout
+        log_file.write(f"Corpus version: {corpus_vers}\n")
+        log_file.write(f"Running on commit {current_commit}\n")
+        log_file.flush()
+
+        # Then, make sure the local corpus is the same as the gcs version
+        reset_to_gcs(t['crate'] + '/corpus/' + t['runner'], REPO_DIR, corpus_vers, log_file)
+        reset_to_gcs(t['crate'] + '/artifacts/' + t['runner'], REPO_DIR, corpus_vers, log_file)
+
+        # Then, make sure we auto-upload any new discovery to gcs
+        stop_inotify_event = threading.Event()
```

Comment:
We cannot, because `stop_inotify_event` also needs to be triggered if the fuzzer finds a crash, but we don't want to stop the entire fuzzing process in this case (it'll just fall back to restarting a fuzzer on another fuzzing target)